lib/src/scanner.dart

```diff
@@ -486,7 +490,12 @@ class Scanner {
       if (key.line == _scanner.line) continue;
 
       if (key.required) {
-        throw YamlException("Expected ':'.", _scanner.emptySpan);
+        if (_recover) {
+          _tokens.insert(key.tokenNumber - _tokensParsed,
+              Token(TokenType.key, key.location.pointSpan() as FileSpan));
```

Comment:
We still need to produce a diagnostic at this point, but then continue to parse the code. In order to do that we'll need to have a way of reporting diagnostics that doesn't involve throwing an exception.

We will eventually want recovery mode to never throw an exception.

Reply:
I've pushed a change that starts this, although I have some questions:

- I added a new `YamlError` class though it's fairly similar to `YamlException` - should the exceptions just be collected instead?
- Should all errors this way (even where we don't currently recover)? For ex. replace all `throw YamlException`s with a call to a function that emits the error and then either recovers or throws (perhaps passing a function that performs the recovery, so that in future it could be made non-nullable to ensure no locations will throw `YamlExceptions` when recovery is enabled).

Reply:
> ... should the exceptions just be collected instead?

I'd probably start with that until / unless we discover a reason to change it.

> Should all errors this way (even where we don't currently recover)?

Eventually the parser should recover from all errors in recovery mode. Some of the errors are probably already things that the parser could safely proceed from, so they should be easy to convert, but proceeding from some of them (without first adding recovery-specific code) would likely produce a large number of useless follow-on diagnostics.

My advice is to continue to throw everywhere that we haven't tested the recover behavior. I'd structure the recovery code like this:

```
if (detectedAnError) {
  reportError(YamlException(...));
  // recover from the error.
}
```
where `reportError` either throws or stores the exception depending on whether we're in recovery mode. That way the recovery code will only be executed in recovery mode and is guaranteed to not slow down the parser when parsing valid input.

The code in server that calls the parser then needs to check the stored exceptions whether one was thrown or not.

> ... perhaps passing a function that performs the recovery ...

I've tried writing parsers with that behavior, but it turns out that it's really hard, in general, to recover well in all cases when operating at a distance like that. I'd recommend against doing it that way and recommend that we keep the recovery code in-line. (The original version of the shared Dart parser was implemented that way, and had to be converted in order to get reasonable recovery.)

Reply:
> >  should the exceptions just be collected instead?
> 
> I'd probably start with that until / unless we discover a reason to change it.

Presumably changing the types of errors later would be a breaking change, though I'm not sure if there would ever be reason to. I'll switch to the exceptions for now - perhaps if the errors were extended in future, it'd be useful to have any new information also on the exception anyway. 

> where `reportError` either throws or stores the exception depending on whether we're in recovery mode.

This is what I started originally, though (if I understand correctly) means `reportError` can't be called if the error doesn't currently have recovery code. I wasn't sure if that was ok - though it sounds like it probably is, so I'll update it.

What I had been thinking about was:

```dart
if (detectedAnError) {
  reportError(YamlException(), () {
    // recover from the error
  });
}
```

The recovery code is still in-line, but in the future making the callback non-null would (hopefully) make it harder to add code reporting new errors (assuming you call `reportError`) without considering recovery. I'm not sure if there's a real benefit there though (given YAML is fairly static, maybe there won't be much code added for new errors in future).